modules/database/database.bal

```diff
@@ -685,9 +707,29 @@ public isolated transactional function getStepsUsingData(int|ExperimentDataFull
 // Timeline ////////////////////////////////////////////////////////////////////
 ////////////////////////////////////////////////////////////////////////////////
 
-public isolated transactional function getTimelineStepCount(int experimentId) returns int|error {
+public isolated transactional function getTimelineStepCount(int experimentId, string? plugin\-name, string? 'version, string? status, int? uncleared\-substep) returns int|error {
+    sql:ParameterizedQuery query = `SELECT count(*) AS rowCount FROM TimelineStep `;
+    if uncleared\-substep != () {
+        query = sql:queryConcat(query, `JOIN TimelineSubstep ON TimelineStep.stepId = TimelineSubstep.stepId `);
```

Comment:
This does not need to compute a full join. An exists clause with a nested select testing for a timelineSubstep with the right stepId and the cleared flag set to false should also work (and may be easier to optimize for the database). Or experiment with querying all stepIds of substeps that are not cleared in a nested query and check if the stepId is in that set.